File: src/adapters/oracle-sqlcl.ts
import { spawn } from "node:child_process";
import { randomUUID } from "node:crypto";
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname } from "node:path";
import { join } from "node:path";
import { BaseSqlAdapter } from "./base-sql.js";
import type { QueryResult } from "../types.js";
import { maskSecret } from "../utils/masking.js";

const SQLCL_META_COMMANDS = new Set([
  "@",
  "@@",
  "accept",
  "append",
  "break",
  "clear",
  "column",
  "connect",
  "copy",
  "define",
  "disconnect",
  "edit",
  "execute",
  "exit",
  "get",
  "host",
  "input",
  "list",
  "password",
  "pause",
  "print",
  "prompt",
  "quit",
  "remark",
  "run",
  "save",
  "set",
  "show",
  "spool",
  "start",
  "undefine",
  "variable",
  "whenever"
]);

export class OracleSqlclAdapter extends BaseSqlAdapter {
  constructor(
    private readonly url: string,
    private readonly sqlclPath = "sql",
    private readonly javaHome?: string
  ) {
    super();
  }

  async connect(): Promise<void> {
    // SQLcl 是短进程 CLI，不维护驱动级长连接；daemon 会复用配置和执行入口。
  }

  async disconnect(): Promise<void> {
    // SQLcl 每次命令执行后进程退出，无需显式断开。
  }

  async execute(command: string): Promise<QueryResult> {
    const result = await this.runSqlcl(command);
    if (result.queryResult) {
      return result.queryResult;
    }
    return {
      rows: [{ output: result.output }],
      fields: ["output"],
      rowCount: result.output ? 1 : 0
    };
  }

  protected testQuery(): string {
    return "select 1 from dual";
  }

  protected async listTables(): Promise<QueryResult> {
    return this.execute("select table_name from user_tables order by table_name");
  }

  protected async listColumns(table: string): Promise<QueryResult> {
    const escaped = table.replace(/'/g, "''").toUpperCase();
    return this.execute(
      `select table_name, column_name, data_type from user_tab_columns where table_name = '${escaped}' order by column_id`
    );
  }

  private runSqlcl(command: string): Promise<{ output: string; queryResult?: QueryResult }> {
    const markers = createQueryResultMarkers();
    return this.withTempScript(command, markers, (scriptPath) => new Promise((resolve, reject) => {
      const child = spawn(this.sqlclPath, ["-S", "/nolog", `@${scriptPath}`], {
        env: this.buildEnv(),
        stdio: ["ignore", "pipe", "pipe"]
      });
      let stdout = "";
      let stderr = "";

      child.stdout.on("data", (chunk) => {
        stdout += chunk.toString("utf8");
      });
      child.stderr.on("data", (chunk) => {
        stderr += chunk.toString("utf8");
      });
      child.on("error", (error) => {
        reject(new Error(`SQLcl 启动失败: ${error.message}`));
      });
      child.on("close", (code) => {
        const stdoutOutput = stripAnsi(stdout).trim();
        const stderrOutput = stripAnsi(stderr).trim();
        const combinedOutput = [stdoutOutput, stderrOutput].filter(Boolean).join("\n");
        if (containsSqlclError(combinedOutput)) {
          reject(new Error(maskSecret(`SQLcl 执行失败: ${combinedOutput}`.trim())));
          return;
        }
        if (code === 0) {
          resolve(parseSqlclOutput(stdoutOutput, stderrOutput, markers));
          return;
        }
        reject(new Error(maskSecret(`SQLcl 执行失败(code=${code}): ${stderr || stdout}`.trim())));
      });
    }));
  }

  private buildScript(command: string, markers: QueryResultMarkers): string {
    const sql = normalizeSqlclSql(command);
    return [
      "set heading on",
      "set feedback off",
      "set pagesize 50000",
      "set linesize 32767",
      "set sqlformat json",
      "whenever sqlerror exit sql.sqlcode",
      `connect ${this.buildConnectString()}`,
      `prompt ${markers.begin}`,
      `${sql};`,
      `prompt ${markers.end}`,
      "exit"
    ].join("\n");
  }

  private buildConnectString(): string {
    const parsed = new URL(this.url);
    const user = decodeURIComponent(parsed.username);
    const password = quotePassword(decodeURIComponent(parsed.password));
    const service = parsed.pathname.replace(/^\//, "");
    return `${user}/${password}@//${parsed.hostname}:${parsed.port || "1521"}/${service}`;
  }

  private buildEnv(): NodeJS.ProcessEnv {
    if (!this.javaHome) {
      return {
        ...process.env,
        NO_COLOR: "1",
        TERM: "dumb"
      };
    }
    const binPath = `${this.javaHome}/bin`;
    return {
      ...process.env,
      JAVA_HOME: this.javaHome,
      PATH: `${binPath}:${dirname(this.sqlclPath)}:${process.env.PATH ?? ""}`,
      NO_COLOR: "1",
      TERM: "dumb"
    };
  }

  private async withTempScript<T>(
    command: string,
    markers: QueryResultMarkers,
    run: (scriptPath: string) => Promise<T>
  ): Promise<T> {
    const directoryPath = await mkdtemp(join(tmpdir(), "agent-database-cli-sqlcl-"));
    const scriptPath = join(directoryPath, "command.sql");
    try {
      await writeFile(scriptPath, this.buildScript(command, markers), { encoding: "utf8", mode: 0o600 });
      return await run(scriptPath);
    } finally {
      await rm(directoryPath, { recursive: true, force: true });
    }
  }
}

function quotePassword(value: string): string {
  if (!/[^\w$#]/.test(value)) {
    return value;
  }
  return `"${value.replace(/"/g, '\\"')}"`;
}

function stripAnsi(value: string): string {
  return value.replace(/\u001b\[[0-9;]*m/g, "");
}

interface QueryResultMarkers {
  begin: string;
  end: string;
}

function createQueryResultMarkers(): QueryResultMarkers {
  const token = randomUUID().replace(/-/g, "");
  return {
    begin: `__DATABASE_CLI_SQLCL_RESULT_BEGIN_${token}__`,
    end: `__DATABASE_CLI_SQLCL_RESULT_END_${token}__`
  };
}

function parseSqlclOutput(
  stdoutOutput: string,
  stderrOutput: string,
  markers: QueryResultMarkers
): { output: string; queryResult?: QueryResult } {
  const preferredOutput = stdoutOutput || stderrOutput;
  const markerResult = extractMarkedSection(stdoutOutput, markers) ?? extractMarkedSection(stderrOutput, markers);
  if (!markerResult) {
    return { output: preferredOutput };
  }
  const parsedJson = tryParseJson(markerResult.section);
  if (!parsedJson) {
    return { output: markerResult.section };
  }
  const rows = normalizeRows(parsedJson);
  if (!rows) {
    return { output: markerResult.section };
  }
  return {
    output: markerResult.section,
    queryResult: {
      rows,
      fields: collectFields(rows),
      rowCount: rows.length
    }
  };
}

function extractMarkedSection(output: string, markers: QueryResultMarkers): { section: string } | undefined {
  const beginIndex = output.indexOf(markers.begin);
  const endIndex = output.indexOf(markers.end);
  if (beginIndex === -1 || endIndex === -1 || endIndex <= beginIndex) {
    return undefined;
  }
  const section = output
    .slice(beginIndex + markers.begin.length, endIndex)
    .trim();
  return { section };
}

function tryParseJson(value: string): unknown | undefined {
  if (!value) {
    return undefined;
  }
  try {
    return JSON.parse(value);
  } catch {
    return undefined;
  }
}

function normalizeRows(value: unknown): unknown[] | undefined {
  if (Array.isArray(value)) {
    return value;
  }
  if (isRecord(value) && Array.isArray(value.items)) {
    return value.items;
  }
  if (
    isRecord(value) &&
    Array.isArray(value.results) &&
    value.results.length > 0 &&
    isRecord(value.results[0]) &&
    Array.isArray(value.results[0].items)
  ) {
    return value.results[0].items;
  }
  return undefined;
}

function collectFields(rows: unknown[]): string[] {
  const fields = new Set<string>();
  for (const row of rows) {
    if (!isRecord(row)) {
      continue;
    }
    for (const key of Object.keys(row)) {
      fields.add(key);
    }
  }
  return [...fields];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function containsSqlclError(output: string): boolean {
  if (!output) {
    return false;
  }
  return /(^|\n)\s*(SP2-|ORA-)/i.test(output);
}

function normalizeSqlclSql(command: string): string {
  const sql = command.trim().replace(/;+\s*$/, "");
  assertSingleSqlStatement(sql);
  assertNoSqlclMetaCommand(sql);
  return sql;
}

function assertSingleSqlStatement(command: string): void {
  const sanitized = stripSqlLiteralsAndComments(command);
  if (sanitized.includes(";")) {
    throw new Error("SQLcl 模式仅允许执行单条 SQL 语句");
  }
}

function assertNoSqlclMetaCommand(command: string): void {
  const sanitized = stripSqlLiteralsAndComments(command);
  for (const line of sanitized.split(/\r?\n/)) {
    const head = line.trim().split(/\s+/)[0]?.toLowerCase();
    if (head && (SQLCL_META_COMMANDS.has(head) || head.startsWith("@"))) {
      throw new Error(`SQLcl 模式拒绝执行 SQLcl 元命令: ${head}`);
    }
  }
}

function stripSqlLiteralsAndComments(command: string): string {
  let result = "";
  let index = 0;

  while (index < command.length) {
    const char = command[index];
    const next = command[index + 1];

    if (char === "-" && next === "-") {
      const endIndex = findLineEnd(command, index + 2);
      result += " ".repeat(endIndex - index);
      index = endIndex;
      continue;
    }

    if (char === "/" && next === "*") {
      const commentEnd = command.indexOf("*/", index + 2);
      const endIndex = commentEnd === -1 ? command.length : commentEnd + 2;
      result += " ".repeat(endIndex - index);
      index = endIndex;
      continue;
    }

    if (char === "'") {
      const endIndex = findQuotedTokenEnd(command, index);
      result += " ".repeat(endIndex - index);
      index = endIndex;
      continue;
    }

    result += char;
    index += 1;
  }

  return result;
}

function findLineEnd(command: string, start: number): number {
  const lineEnd = command.indexOf("\n", start);
  return lineEnd === -1 ? command.length : lineEnd;
}

function findQuotedTokenEnd(command: string, start: number): number {
  let index = start + 1;
  while (index < command.length) {
    if (command[index] === "'") {
      if (command[index + 1] === "'") {
        index += 2;
        continue;
      }
      return index + 1;
    }
    index += 1;
  }
  return command.length;
}

File: rust/src/adapters/oracle_sqlcl.rs
use super::DatabaseAdapter;
use crate::types::{MetadataRequest, MetadataType, QueryResult};
use crate::utils::masking::mask_secret;
use anyhow::Result;
use async_trait::async_trait;
use serde_json::Value;
use std::{fs, path::PathBuf, process::Command};
use url::Url;
use uuid::Uuid;

pub struct OracleSqlclAdapter {
    url: String,
    sqlcl_path: String,
    java_home: Option<String>,
}

impl OracleSqlclAdapter {
    pub fn new(url: String, sqlcl_path: String, java_home: Option<String>) -> Self {
        Self {
            url,
            sqlcl_path,
            java_home,
        }
    }

    fn run_sqlcl(&self, command: &str) -> Result<QueryResult> {
        let markers = Markers::new();
        let dir = std::env::temp_dir().join(format!("agent-database-cli-sqlcl-{}", Uuid::new_v4()));
        fs::create_dir_all(&dir)?;
        let script_path = dir.join("command.sql");
        fs::write(&script_path, self.build_script(command, &markers))?;
        let result = self.spawn_sqlcl(&script_path);
        let _ = fs::remove_dir_all(&dir);
        result
    }

    fn spawn_sqlcl(&self, script_path: &PathBuf) -> Result<QueryResult> {
        let mut command = Command::new(&self.sqlcl_path);
        command
            .arg("-S")
            .arg("/nolog")
            .arg(format!("@{}", script_path.display()))
            .env("NO_COLOR", "1")
            .env("TERM", "dumb");
        if let Some(java_home) = &self.java_home {
            command.env("JAVA_HOME", java_home);
        }
        let output = command
            .output()
            .map_err(|error| anyhow::anyhow!("SQLcl 启动失败: {}", error))?;
        let stdout = strip_ansi(&String::from_utf8_lossy(&output.stdout));
        let stderr = strip_ansi(&String::from_utf8_lossy(&output.stderr));
        let combined = [stdout.trim(), stderr.trim()]
            .into_iter()
            .filter(|v| !v.is_empty())
            .collect::<Vec<_>>()
            .join("\n");
        if contains_sqlcl_error(&combined) || !output.status.success() {
            anyhow::bail!("{}", mask_secret(&format!("SQLcl 执行失败: {combined}")));
        }
        parse_sqlcl_output(&stdout)
    }

    fn build_script(&self, command: &str, markers: &Markers) -> String {
        format!("set heading on\nset feedback off\nset pagesize 50000\nset linesize 32767\nset sqlformat json\nwhenever sqlerror exit sql.sqlcode\nconnect {}\nprompt {}\n{};\nprompt {}\nexit", self.build_connect_string(), markers.begin, normalize_sqlcl_sql(command), markers.end)
    }

    fn build_connect_string(&self) -> String {
        let parsed = Url::parse(&self.url).expect("Oracle URL 已在配置阶段校验");
        let user = parsed.username();
        let password = quote_password(parsed.password().unwrap_or(""));
        let service = parsed.path().trim_start_matches('/');
        format!(
            "{user}/{password}@//{}:{}/{}",
            parsed.host_str().unwrap_or("localhost"),
            parsed.port().unwrap_or(1521),
            service
        )
    }
}

#[async_trait]
impl DatabaseAdapter for OracleSqlclAdapter {
    async fn connect(&mut self) -> Result<()> {
        Ok(())
    }
    async fn disconnect(&mut self) -> Result<()> {
        Ok(())
    }
    async fn test(&mut self) -> Result<()> {
        self.execute("select 1 from dual").await.map(|_| ())
    }
    async fn execute(&mut self, command: &str) -> Result<QueryResult> {
        self.run_sqlcl(command)
    }
    async fn metadata(&mut self, request: MetadataRequest) -> Result<QueryResult> {
        match request.request_type {
            MetadataType::Tables => {
                self.execute("select table_name from user_tables order by table_name")
                    .await
            }
            MetadataType::Columns => {
                let table = request
                    .table
                    .ok_or_else(|| anyhow::anyhow!("columns 元信息查询必须提供 --table"))?
                    .replace('\'', "''")
                    .to_uppercase();
                self.execute(&format!("select table_name, column_name, data_type from user_tab_columns where table_name = '{}' order by column_id", table)).await
            }
            _ => anyhow::bail!("当前数据库不支持元信息类型: {:?}", request.request_type),
        }
    }
}

struct Markers {
    begin: String,
    end: String,
}
impl Markers {
    fn new() -> Self {
        let id = Uuid::new_v4();
        Self {
            begin: format!("__AGENT_DATABASE_CLI_BEGIN_{id}__"),
            end: format!("__AGENT_DATABASE_CLI_END_{id}__"),
        }
    }
}
fn normalize_sqlcl_sql(command: &str) -> String {
    command.trim().trim_end_matches(';').to_string()
}
fn quote_password(password: &str) -> String {
    format!("\"{}\"", password.replace('"', "\\\""))
}
fn strip_ansi(value: &str) -> String {
    regex::Regex::new(r"\x1b\[[0-9;]*m")
        .unwrap()
        .replace_all(value, "")
        .to_string()
}
fn contains_sqlcl_error(value: &str) -> bool {
    ["ORA-", "SP2-", "SQL Error"]
        .iter()
        .any(|item| value.contains(item))
}
fn parse_sqlcl_output(stdout: &str) -> Result<QueryResult> {
    let json_start = stdout.find('[').or_else(|| stdout.find('{'));
    if let Some(start) = json_start {
        let slice = &stdout[start..];
        if let Ok(value) = serde_json::from_str::<Value>(slice.trim()) {
            let rows = value.as_array().cloned().unwrap_or_else(|| vec![value]);
            return Ok(QueryResult {
                row_count: Some(rows.len() as u64),
                rows,
                fields: None,
            });
        }
    }
    Ok(QueryResult {
        rows: vec![serde_json::json!({ "output": stdout.trim() })],
        fields: Some(vec!["output".to_string()]),
        row_count: if stdout.trim().is_empty() {
            Some(0)
        } else {
            Some(1)
        },
    })
}
